File: src/main/java/dev/antigravity/mazeward/core/PathFinder.java
package dev.antigravity.mazeward.core;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

/**
 * 任意角度経路探索（Theta*）。
 *
 * <p>グリッド A* だと敵が「マス目に沿ってカクカク」動いてしまう。
 * このゲームは敵の経路そのものを見せるのが売りなので、
 * <b>角から角へ直線で、45°に限らない任意の角度で</b> 移動させたい。
 * そのために A* ではなく Theta* を使っている。</p>
 *
 * <p>Theta* は展開のたびに「親から直接見通せるか（line of sight）」を判定し、
 * 見通せるなら中間ノードを飛ばして親を引き継ぐ。結果として
 * <b>曲がり角は必ず「障害物に接している通行可能セル」の中心になる</b>。
 * 障害物の角そのものではなく、その隣の空きセルの中心が折れ点になる。</p>
 *
 * <pre>
 *   A* (4方向)              Theta* (任意角度)
 *   S─┐                     S
 *     │                      ＼
 *     └─┐                      ＼
 *       │                        ＼
 *       └─ G                       G
 * </pre>
 *
 * <h2>決定性</h2>
 * 配置プレビューの信頼性のため、同じ盤面からは必ず同じ経路が返る必要がある。
 * そのために近傍の展開順を固定し、優先度が同値のときは h → 挿入順で決着させている。
 *
 * <h2>角抜けの禁止</h2>
 * 斜め移動と見通し判定はどちらも「障害物の角を斜めにすり抜けない」ルールで統一している。
 * 敵の当たり判定には幅があるので、角をかすめる経路は壁にめり込んで見えてしまう。
 * このため、斜めに進むには両隣のセルが空いている必要がある。
 * 結果として、壁の角を回るときは必ずその隣の空きセル（＝曲がり角）を経由する。
 */
public final class PathFinder {

    /** 北 → 北東 → 東 → 南東 → 南 → 南西 → 西 → 北西。展開順を変えると経路の見た目が変わるので固定。 */
    private static final int[] DX = {0, 1, 1, 1, 0, -1, -1, -1};
    private static final int[] DZ = {-1, -1, 0, 1, 1, 1, 0, -1};

    private static final double EPSILON = 1e-9;

    private PathFinder() {
    }

    public static PathResult find(Grid grid, Vec2i start, Vec2i goal) {
        return find(grid, start, List.of(goal));
    }

    public static PathResult find(Grid grid, Vec2i start, Collection<Vec2i> goals) {
        if (!grid.inBounds(start) || !grid.walkable(start) || goals.isEmpty()) {
            return PathResult.unreachable();
        }

        int width = grid.width();
        int height = grid.height();
        int cellCount = width * height;

        boolean[] isGoal = new boolean[cellCount];
        List<Vec2i> goalList = new ArrayList<>(goals.size());
        for (Vec2i goal : goals) {
            if (grid.inBounds(goal)) {
                isGoal[goal.z() * width + goal.x()] = true;
                goalList.add(goal);
            }
        }
        if (goalList.isEmpty()) {
            return PathResult.unreachable();
        }

        double[] gScore = new double[cellCount];
        Arrays.fill(gScore, Double.POSITIVE_INFINITY);
        int[] parent = new int[cellCount];
        Arrays.fill(parent, -1);
        boolean[] closed = new boolean[cellCount];

        int startIndex = start.z() * width + start.x();
        gScore[startIndex] = 0.0;
        parent[startIndex] = startIndex;

        PriorityQueue<Node> open = new PriorityQueue<>(
                Comparator.comparingDouble(Node::f)
                        .thenComparingDouble(Node::h)
                        .thenComparingLong(Node::seq));
        long sequence = 0;
        double startH = heuristic(start.x(), start.z(), goalList);
        open.add(new Node(startIndex, startH, startH, sequence++));

        while (!open.isEmpty()) {
            Node current = open.poll();
            int index = current.index();
            if (closed[index]) {
                continue;
            }
            closed[index] = true;

            if (isGoal[index]) {
                return rebuild(parent, index, width, gScore[index]);
            }

            int cx = index % width;
            int cz = index / width;
            int parentIndex = parent[index];
            int px = parentIndex % width;
            int pz = parentIndex / width;

            for (int direction = 0; direction < 8; direction++) {
                int dx = DX[direction];
                int dz = DZ[direction];
                if (!grid.canStep(cx, cz, dx, dz)) {
                    continue;
                }
                int nx = cx + dx;
                int nz = cz + dz;
                int neighbour = nz * width + nx;
                if (closed[neighbour]) {
                    continue;
                }

                // Theta* の肝: 親から直接見通せるなら、いま展開しているノードを飛ばして
                // 親から直線でつなぐ。これで経路が任意角度の直線になる。
                double tentative;
                int newParent;
                if (lineOfSight(grid, px, pz, nx, nz)) {
                    tentative = gScore[parentIndex] + distance(px, pz, nx, nz);
                    newParent = parentIndex;
                } else {
                    tentative = gScore[index] + distance(cx, cz, nx, nz);
                    newParent = index;
                }

                if (tentative < gScore[neighbour] - EPSILON) {
                    gScore[neighbour] = tentative;
                    parent[neighbour] = newParent;
                    double h = heuristic(nx, nz, goalList);
                    open.add(new Node(neighbour, tentative + h, h, sequence++));
                }
            }
        }

        return PathResult.unreachable();
    }

    /**
     * 2 つのセルの中心を結ぶ直線が、通行可能セルだけを通るか。
     *
     * <p>線分が触れるセルをすべて拾う「スーパーカバー」で走査する。
     * 線分がちょうど格子点（4 セルの角）を通るときは、
     * 斜めの隙間をすり抜けたことになるので、両隣が空いている場合だけ許可する。
     * これがないと、敵が壁の角にめり込んで見える。</p>
     */
    public static boolean lineOfSight(Grid grid, int x0, int z0, int x1, int z1) {
        if (!grid.walkable(x0, z0) || !grid.walkable(x1, z1)) {
            return false;
        }
        if (x0 == x1 && z0 == z1) {
            return true;
        }

        int dx = x1 - x0;
        int dz = z1 - z0;
        int stepX = Integer.signum(dx);
        int stepZ = Integer.signum(dz);
        int absX = Math.abs(dx);
        int absZ = Math.abs(dz);

        int x = x0;
        int z = z0;
        int error = absX - absZ;
        int doubleX = absX * 2;
        int doubleZ = absZ * 2;

        for (int remaining = 1 + absX + absZ; remaining > 0; remaining--) {
            if (!grid.walkable(x, z)) {
                return false;
            }
            if (error > 0) {
                x += stepX;
                error -= doubleZ;
            } else if (error < 0) {
                z += stepZ;
                error += doubleX;
            } else {
                // 格子点をちょうど通過する = 斜めに角を抜ける。両隣が空いていなければ不可。
                if (!grid.walkable(x + stepX, z) || !grid.walkable(x, z + stepZ)) {
                    return false;
                }
                x += stepX;
                z += stepZ;
                error -= doubleZ;
                error += doubleX;
                remaining--;
            }
        }
        return true;
    }

    public static boolean lineOfSight(Grid grid, Vec2i from, Vec2i to) {
        return lineOfSight(grid, from.x(), from.z(), to.x(), to.z());
    }

    /**
     * 折れ線が実際に通過するセルの一覧。
     * 描画やタワー配置の補助（経路の近くか判定する）で使う。
     */
    public static List<Vec2i> traversedCells(List<Vec2i> waypoints) {
        List<Vec2i> cells = new ArrayList<>();
        if (waypoints.isEmpty()) {
            return cells;
        }
        cells.add(waypoints.get(0));
        for (int i = 1; i < waypoints.size(); i++) {
            appendSegmentCells(cells, waypoints.get(i - 1), waypoints.get(i));
        }
        return cells;
    }

    private static void appendSegmentCells(List<Vec2i> out, Vec2i from, Vec2i to) {
        int stepX = Integer.signum(to.x() - from.x());
        int stepZ = Integer.signum(to.z() - from.z());
        int absX = Math.abs(to.x() - from.x());
        int absZ = Math.abs(to.z() - from.z());

        int x = from.x();
        int z = from.z();
        int error = absX - absZ;
        int doubleX = absX * 2;
        int doubleZ = absZ * 2;

        for (int remaining = 1 + absX + absZ; remaining > 0; remaining--) {
            Vec2i cell = new Vec2i(x, z);
            if (out.isEmpty() || !out.get(out.size() - 1).equals(cell)) {
                out.add(cell);
            }
            if (absX == 0 && absZ == 0) {
                return;
            }
            if (error > 0) {
                x += stepX;
                error -= doubleZ;
            } else if (error < 0) {
                z += stepZ;
                error += doubleX;
            } else {
                x += stepX;
                z += stepZ;
                error -= doubleZ;
                error += doubleX;
                remaining--;
            }
        }
    }

    /**
     * 2 つの折れ線を比べて <b>変化した区間だけ</b> を返す。
     *
     * <p>障害物を置いたときの予測経路は、実際には経路の一部しか変わらないことが多い。
     * 全体を赤で描くと「どこが変わったのか」がかえって読めなくなるので、
     * 前後の共通部分を取り除いて、変化した区間だけを描けるようにする。</p>
     *
     * <p>つなぎ目が浮かないよう、変化部分の前後 1 点ずつは残す。
     * 完全に同じなら空リスト（＝赤を一切描かない）。</p>
     */
    public static List<Vec2i> divergentSection(List<Vec2i> before, List<Vec2i> after) {
        if (after.isEmpty() || before.equals(after)) {
            return List.of();
        }
        int beforeSize = before.size();
        int afterSize = after.size();

        int prefix = 0;
        while (prefix < beforeSize && prefix < afterSize
                && before.get(prefix).equals(after.get(prefix))) {
            prefix++;
        }
        int suffix = 0;
        while (suffix < beforeSize - prefix && suffix < afterSize - prefix
                && before.get(beforeSize - 1 - suffix).equals(after.get(afterSize - 1 - suffix))) {
            suffix++;
        }

        int from = Math.max(0, prefix - 1);
        int to = Math.min(afterSize, afterSize - suffix + 1);
        if (to <= from) {
            return List.of();
        }
        return List.copyOf(after.subList(from, to));
    }

    /** 折れ線の全長（セル中心を結んだユークリッド距離）。 */
    public static double polylineLength(List<Vec2i> waypoints) {
        double total = 0;
        for (int i = 1; i < waypoints.size(); i++) {
            Vec2i a = waypoints.get(i - 1);
            Vec2i b = waypoints.get(i);
            total += distance(a.x(), a.z(), b.x(), b.z());
        }
        return total;
    }

    private static PathResult rebuild(int[] parent, int goalIndex, int width, double length) {
        List<Vec2i> reversed = new ArrayList<>();
        int index = goalIndex;
        while (true) {
            reversed.add(new Vec2i(index % width, index / width));
            int next = parent[index];
            if (next == index || next < 0) {
                break;
            }
            index = next;
        }
        List<Vec2i> waypoints = new ArrayList<>(reversed.size());
        for (int i = reversed.size() - 1; i >= 0; i--) {
            waypoints.add(reversed.get(i));
        }
        return new PathResult(List.copyOf(waypoints), length, true);
    }

    private static double distance(int x0, int z0, int x1, int z1) {
        double dx = x1 - x0;
        double dz = z1 - z0;
        return Math.sqrt(dx * dx + dz * dz);
    }

    private static double heuristic(int x, int z, List<Vec2i> goals) {
        double best = Double.POSITIVE_INFINITY;
        for (Vec2i goal : goals) {
            double dx = x - goal.x();
            double dz = z - goal.z();
            double d = Math.sqrt(dx * dx + dz * dz);
            if (d < best) {
                best = d;
            }
        }
        return best;
    }

    private record Node(int index, double f, double h, long seq) {
    }
}

File: ai/mazeward_env/pathfinder.py
# -*- coding: utf-8 -*-
"""任意角度経路探索 Theta*。 ``core/PathFinder.java`` の移植。

グリッド A* だと敵がマス目に沿ってカクカク動く。このゲームは
**曲がり角が「障害物に接した通行可能セルの中心」になる任意角度の折れ線**
であることが迷路設計の前提なので、A* ではなく Theta* を使う。
迷路の「長さ」ではなく「形」が効くのはこの性質のおかげ。

移植で守ったこと
----------------
- **展開順を固定**（北→北東→東→…）。順序を変えると同じ盤面から違う経路が出る
- **同値の決着を f → h → 挿入順**にする。プレビューの信頼性の根拠が決定性なので
- **角抜けの禁止**を斜め移動と見通し判定の両方で統一する

ここを numpy でベクトル化しなかったのは、優先度付きキューが本質的に逐次だから。
代わりに「**壁を置いたときだけ**呼ぶ」という Java 側と同じ設計にして、
呼ばれる回数のほうを減らしてある（毎 tick 全敵ぶん引き直すことは絶対にしない）。
"""

from __future__ import annotations

import heapq
import math
from typing import List, Optional, Sequence, Tuple

import numpy as np

from .grid import Grid

DX = (0, 1, 1, 1, 0, -1, -1, -1)
DZ = (-1, -1, 0, 1, 1, 1, 0, -1)
EPSILON = 1e-9


class PathResult:
    """折れ線と実距離。 (core/PathResult.java)

    :param waypoints: 曲がり角のセル座標。実際の通過点はセル中心 (x+0.5, z+0.5)
    :param length: ユークリッド長（マス数ではない）
    """

    __slots__ = ("waypoints", "length", "reachable")

    def __init__(self, waypoints: List[Tuple[int, int]], length: float,
                 reachable: bool):
        self.waypoints = waypoints
        self.length = length
        self.reachable = reachable

    @staticmethod
    def unreachable() -> "PathResult":
        return PathResult([], 0.0, False)

    def turns(self) -> int:
        return max(0, len(self.waypoints) - 2)


_UNREACHABLE = PathResult([], 0.0, False)


def line_of_sight(walk: np.ndarray, x0: int, z0: int, x1: int, z1: int) -> bool:
    """2 セル中心を結ぶ直線が通行可能セルだけを通るか。 (PathFinder#lineOfSight)

    線分が触れるセルをすべて拾う「スーパーカバー」で走査する。ちょうど格子点
    （4 セルの角）を通るときは斜めに角を抜けたことになるので、両隣が空いている
    場合だけ許す。これがないと敵が壁の角にめり込んで見える。
    """
    h, w = walk.shape
    if not (0 <= x0 < w and 0 <= z0 < h and 0 <= x1 < w and 0 <= z1 < h):
        return False
    if not walk[z0, x0] or not walk[z1, x1]:
        return False
    if x0 == x1 and z0 == z1:
        return True

    dx, dz = x1 - x0, z1 - z0
    step_x = (dx > 0) - (dx < 0)
    step_z = (dz > 0) - (dz < 0)
    abs_x, abs_z = abs(dx), abs(dz)

    x, z = x0, z0
    error = abs_x - abs_z
    double_x, double_z = abs_x * 2, abs_z * 2

    remaining = 1 + abs_x + abs_z
    while remaining > 0:
        remaining -= 1
        if not (0 <= x < w and 0 <= z < h) or not walk[z, x]:
            return False
        if error > 0:
            x += step_x
            error -= double_z
        elif error < 0:
            z += step_z
            error += double_x
        else:
            # 格子点をちょうど通過 = 斜めに角を抜ける。両隣が空いていなければ不可。
            # 盤外は Java の Grid#get と同じく BORDER（通行不可）として扱う
            nx, nz = x + step_x, z + step_z
            if not (0 <= nx < w and walk[z, nx]):
                return False
            if not (0 <= nz < h and walk[nz, x]):
                return False
            x += step_x
            z += step_z
            error += double_x - double_z
            remaining -= 1
    return True


def find(grid: Grid, start: Tuple[int, int],
         goals: Optional[Sequence[Tuple[int, int]]] = None) -> PathResult:
    """スポーンからコアまでの折れ線。 (PathFinder#find)"""
    goals = grid.core_cells if goals is None else goals
    sx, sz = start
    if not grid.in_bounds(sx, sz) or not goals:
        return _UNREACHABLE
    walk = grid.walk
    if not walk[sz, sx]:
        return _UNREACHABLE

    w, h = grid.width, grid.height
    cell_count = w * h

    is_goal = np.zeros(cell_count, dtype=bool)
    goal_list: List[Tuple[int, int]] = []
    for gx, gz in goals:
        if grid.in_bounds(gx, gz):
            is_goal[gz * w + gx] = True
            goal_list.append((gx, gz))
    if not goal_list:
        return _UNREACHABLE

    g_score = [math.inf] * cell_count
    parent = [-1] * cell_count
    closed = bytearray(cell_count)

    def heuristic(x: int, z: int) -> float:
        best = math.inf
        for gx, gz in goal_list:
            d = math.hypot(x - gx, z - gz)
            if d < best:
                best = d
        return best

    start_index = sz * w + sx
    g_score[start_index] = 0.0
    parent[start_index] = start_index

    seq = 0
    start_h = heuristic(sx, sz)
    open_heap: List[Tuple[float, float, int, int]] = [(start_h, start_h, seq, start_index)]
    seq += 1

    while open_heap:
        _, _, _, index = heapq.heappop(open_heap)
        if closed[index]:
            continue
        closed[index] = 1

        if is_goal[index]:
            return _rebuild(parent, index, w, g_score[index])

        cx, cz = index % w, index // w
        parent_index = parent[index]
        px, pz = parent_index % w, parent_index // w
        g_parent = g_score[parent_index]
        g_current = g_score[index]

        for d in range(8):
            dx, dz = DX[d], DZ[d]
            nx, nz = cx + dx, cz + dz
            if nx < 0 or nz < 0 or nx >= w or nz >= h or not walk[nz, nx]:
                continue
            if dx and dz and not (walk[cz, nx] and walk[nz, cx]):
                continue
            neighbour = nz * w + nx
            if closed[neighbour]:
                continue

            # Theta* の肝: 親から直接見通せるなら、いま展開しているノードを
            # 飛ばして親から直線でつなぐ。結果が任意角度の直線になる
            if line_of_sight(walk, px, pz, nx, nz):
                tentative = g_parent + math.hypot(nx - px, nz - pz)
                new_parent = parent_index
            else:
                tentative = g_current + math.hypot(dx, dz)
                new_parent = index

            if tentative < g_score[neighbour] - EPSILON:
                g_score[neighbour] = tentative
                parent[neighbour] = new_parent
                hh = heuristic(nx, nz)
                heapq.heappush(open_heap, (tentative + hh, hh, seq, neighbour))
                seq += 1

    return _UNREACHABLE


def _rebuild(parent: List[int], goal_index: int, width: int,
             length: float) -> PathResult:
    reversed_cells: List[Tuple[int, int]] = []
    index = goal_index
    while True:
        reversed_cells.append((index % width, index // width))
        nxt = parent[index]
        if nxt == index or nxt < 0:
            break
        index = nxt
    reversed_cells.reverse()
    return PathResult(reversed_cells, length, True)


def polyline_length(waypoints: Sequence[Tuple[int, int]]) -> float:
    """折れ線の全長。 (PathFinder#polylineLength)"""
    total = 0.0
    for i in range(1, len(waypoints)):
        ax, az = waypoints[i - 1]
        bx, bz = waypoints[i]
        total += math.hypot(bx - ax, bz - az)
    return total


def traversed_cells(waypoints: Sequence[Tuple[int, int]]) -> List[Tuple[int, int]]:
    """折れ線が実際に通過するセル。経路のラスタライズ（観測チャンネル）に使う。
    (PathFinder#traversedCells)"""
    cells: List[Tuple[int, int]] = []
    if not waypoints:
        return cells
    cells.append(tuple(waypoints[0]))
    for i in range(1, len(waypoints)):
        _append_segment(cells, waypoints[i - 1], waypoints[i])
    return cells


def _append_segment(out: List[Tuple[int, int]], a: Tuple[int, int],
                    b: Tuple[int, int]) -> None:
    step_x = (b[0] > a[0]) - (b[0] < a[0])
    step_z = (b[1] > a[1]) - (b[1] < a[1])
    abs_x, abs_z = abs(b[0] - a[0]), abs(b[1] - a[1])
    x, z = a
    error = abs_x - abs_z
    double_x, double_z = abs_x * 2, abs_z * 2
    remaining = 1 + abs_x + abs_z
    while remaining > 0:
        remaining -= 1
        if not out or out[-1] != (x, z):
            out.append((x, z))
        if abs_x == 0 and abs_z == 0:
            return
        if error > 0:
            x += step_x
            error -= double_z
        elif error < 0:
            z += step_z
            error += double_x
        else:
            x += step_x
            z += step_z
            error += double_x - double_z
            remaining -= 1


def to_points(waypoints: Sequence[Tuple[int, int]]) -> np.ndarray:
    """セル座標 → 通過点（セル中心）の (n, 2) float 配列。"""
    if not waypoints:
        return np.zeros((0, 2), dtype=np.float32)
    arr = np.asarray(waypoints, dtype=np.float32) + 0.5
    return arr


def cumulative(points: np.ndarray) -> np.ndarray:
    """折れ線の各点までの累積距離 (n,)。 (EnemyInstance#applyWaypoints)"""
    if len(points) == 0:
        return np.zeros(0, dtype=np.float32)
    seg = np.linalg.norm(np.diff(points, axis=0), axis=1)
    out = np.zeros(len(points), dtype=np.float32)
    out[1:] = np.cumsum(seg)
    return out
